content/posts/2024-08-29-GSoC-2024-statistical_analysis_of_llvm-ir_compilation.md

```diff
@@ -0,0 +1,101 @@
+---
+author: "Andrew Kallai"
+date: "2024-08-29"
+tags: ["GSoC", "llvm-ir-dataset-utils", "ComPile", "analysis"]
+title: "GSoC 2024: Statistical Analysis of LLVM-IR Compilation"
+---
+
+Welcome! My name is Andrew and I contributed to LLVM through the 2024 Google Summer of Code Program. My project is called [Statistical Analysis of LLVM-IR Compilation](https://summerofcode.withgoogle.com/programs/2024/projects/hquDyVBK). The objective of this project is to provide an analysis of the causes of unexpected compilation times.
+
+# Background
+
+In principle, an LLVM IR bitcode file, or module, contains IR features that determine the behavior of the compiler optimization pipeline. By varying these features, the optimization pipeline, opt, can perform better or worse. More specifically, optimizations succeed in less or more time; the user can wait for a microsecond or a few minutes. LLVM compiler developers constantly edit the pipeline, so the performance of these optimizations can vary by time (sometimes by seconds).
+
+Having a large IR dataset such as ComPile (linked below) allows for testing the LLVM compilation pipeline on a varied sample of IR. The size of this sample is sufficient to determine outlying IR modules. By identifying and examining such files using utilities which are being added to the LLVM IR Dataset Utils Repo (linked below), the causes of unexpected compilation times can be determined. Developers can then modify and improve the compilation pipeline accordingly.
+
+# Summary of Work
+
+The utilities added in PR37 (linked below) are intended to write each IR module to a tar file corresponding to a programming language. This is designed to be a multithreaded process that is agnostic to the number of files per language or the type of language. Executing the scripts correctly results in a file containing index values. Each file written to the tar files is indexed by its location in the HF dataset. The first file index begins at 1, so for N files, there are IR files from file1.bc to fileN.bc. 
```

Comment:
You can probably insert direct link here as well. And in few other places where you're referencing a PR.